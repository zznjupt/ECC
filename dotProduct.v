module dotProduct # (
    parameter DATA_WIDTH = 256
) (
    input  wire                     clk,
    input  wire                     rst_n,
    input  wire [DATA_WIDTH-1:0]    Px,
    input  wire [DATA_WIDTH-1:0]    Py,
    input  wire [DATA_WIDTH-1:0]    k,
    output reg  [DATA_WIDTH-1:0]    Rx,
    output reg  [DATA_WIDTH-1:0]    Ry,
    input  wire                     in_valid,
    output reg                      out_valid
);


parameter IDLE  = 4'd0;
parameter RD	= 4'd1;
parameter IDLE2	= 4'd2;
parameter ADD1	= 4'd3;
parameter IDLE3	= 4'd4;
parameter ADD2 	= 4'd5;
parameter IDLE4	= 4'd6;
parameter DONE	= 4'd7;

reg [3:0] state_cs, state_ns;
reg [9:0] cnt;

reg [DATA_WIDTH-1:0] Px_reg;
reg [DATA_WIDTH-1:0] Py_reg;
reg [DATA_WIDTH-1:0] k_reg;

wire done_add;
wire add_Mult;
assign add_Mult = (k_reg[cnt]) ? 1 : 0;
assign done_add = (cnt == DATA_WIDTH) ? 1 : 0;

reg  [DATA_WIDTH-1:0]   add_Px;
reg  [DATA_WIDTH-1:0]   add_Py;
reg  [DATA_WIDTH-1:0]   add_Qx;
reg  [DATA_WIDTH-1:0]   add_Qy;
wire [DATA_WIDTH-1:0]   add_Rx;
wire [DATA_WIDTH-1:0]   add_Ry;
reg 		            add_in_valid;
wire 		            add_out_valid;

add # (.DATA_WIDTH(DATA_WIDTH)) 
a1 (
	.clk        (clk),
	.rst_n      (rst_n),
	.Px         (add_Px),
	.Py         (add_Py),
	.Qx         (add_Qx),
	.Qy         (add_Qy),
	.Rx         (add_Rx),
	.Ry         (add_Ry),
	.in_valid   (add_in_valid),
	.out_valid  (add_out_valid)
);

always @(*) begin
	if((state_cs != ADD1 && state_ns == ADD1) || (state_cs != ADD2 && state_ns == ADD2))
		add_in_valid = 1;
	else
		add_in_valid = 0;
end

always @(*) begin
	if(state_ns == ADD1)begin
		add_Px = Px_reg;
		add_Py = Py_reg;
		add_Qx = Rx;
		add_Qy = Ry;
	end else if(state_ns == ADD2) begin
		add_Px = Px_reg;
		add_Py = Py_reg;
		add_Qx = Px_reg;
		add_Qy = Py_reg;
	end else begin
		add_Px = 0;
		add_Py = 0;
		add_Qx = 0;
		add_Qy = 0;
	end
end

always @(posedge clk or negedge rst_n) begin
	if(!rst_n)
		state_cs <= IDLE;
	else
		state_cs <= state_ns;
end

always @(*) begin
	case(state_cs)
		IDLE:	state_ns = (in_valid) ? RD:IDLE;
		RD:		state_ns = IDLE2;
		IDLE2:	begin
			if(done_add)
				state_ns = IDLE4;
			else if(add_Mult)
				state_ns = ADD1;
			else
				state_ns = ADD2;
		end
		ADD1:	state_ns = (add_out_valid) ? IDLE3 : ADD1;
		IDLE3:	state_ns = ADD2;
		ADD2:	state_ns = (add_out_valid) ? IDLE2 : ADD2;
		IDLE4:	state_ns = DONE;
		DONE:	state_ns = IDLE;
		// default:state_ns = IDLE;
	endcase
end

always @(posedge clk or negedge rst_n) begin
	if(!rst_n)
		cnt <= 0;
	else if(state_ns == IDLE)
		cnt <= 0;
	else if(state_cs == ADD2 && state_ns == IDLE2)
		cnt <= cnt+1;
end

// register
always @(posedge clk or negedge rst_n) begin
	if(!rst_n) begin
		Px_reg <= 0;
		Py_reg <= 0;
		k_reg <= 0;
	end else if(state_ns == IDLE) begin
		Px_reg <= 0;
		Py_reg <= 0;
		k_reg <= 0;
	end else if(state_cs == ADD2 && add_out_valid) begin
		Px_reg <= add_Rx;
		Py_reg <= add_Ry;
	end else if(state_ns == RD) begin
		Px_reg <= Px;
		Py_reg <= Py;
		k_reg <= k;
	end
end

// output
always @(posedge clk or negedge rst_n) begin
	if(!rst_n) begin
		out_valid <= 0;
		Rx <= 0;
		Ry <= 0;
	end else if(state_ns == IDLE) begin
		out_valid <= 0;
		Rx <= 0;
		Ry <= 0;
	end else if(state_cs == ADD1 && add_out_valid) begin
		Rx <= add_Rx;
		Ry <= add_Ry;
	end else if(state_ns == DONE) begin
		out_valid <= 1;		
	end
end

endmodule