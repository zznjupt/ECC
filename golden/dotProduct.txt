79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
EBB2C082FD7727890A28AC82F6BDF97BAD8DE9F5D7C90286
be68fc452d7daf02f8135e790a9508768cb0f9a63e9d6651
6317d91607e520c4d446f105c74b319087f75f18cb04f25a

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
49A0D7B786EC9CDE0D0721D72804BEFD06571C974B191EFB
7551a3e0c48b19d805dc6815abcec77d73fc871db5466826
42df45c5076eba4f287994cecba79606773991af19628bd4

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03B
e6685dec81abd3e9969a5037319e7474a7298296fd759a2b
1dc055ffa1f0bb58e232a11e62376c609c1430294ffb10a9

79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D9
483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A6855419
79F6B20A993753B8E44FB9EB86617FA217A759026549F68F
b8c8a3f548eee5aa3683ea60cb46f695d339bc5e9d2ea5bd
6ad7551198d3179a8d5830b5a57f73375215d520589f1ebd

779DD197A5DF977ED2CF6CB31D82D43328B790DC6B3B7D44
E94B724A555B6D017BB7607C3E3281DAF5B1699D6EF41249
3E78FC6B73EB6865D559960AF25A2E6303A7DB0CD418A4F0
193b7e95316644db6255f0f3d43948c1ba834f49cf4ff3bf
a88a47f66ff2f24920948a28038755f1d47f3e721fd89a47